### DISCUSSION TITLE
About getting the current time.

### DISCUSSION THREAD

**Discussion by cj2a7t** [Kong/kong#11552] [Help] [answered]
In openresty or lua, what is the best way to get the current time unit in microseconds?

**Comment by chronolaw** [comment, marked as answer]
OpenResty is based on Nginx, and Nginx normally uses millisecond to measure time.

So if you want to use microsecond perhaps you have to write some ffi code to call C functions.

**Comment by chronolaw** [comment]
This link may be helpful: https://github.com/Kong/kong/blob/ddc1400fca7f8b1cf7bec46c8970da619ef5a9ab/kong/tools/utils.lua#L1684-L1695

**Reply by cj2a7t** [reply]
How is the performance of this method? @chronolaw

**Reply by oowl** [reply]
It can not catch up `ngx.now()` performance, Because `clock_gettime` will do syscall every function call. But `ngx.now()` related time is getting time from the nginx time cache,

**Reply by cj2a7t** [reply]
@oowl   How does it compare to ngx.update_time + ngx.now()

**Reply by oowl** [reply]
IMO, they have similar performance indicators. But we have not tested it.

**Reply by cj2a7t** [reply]
This requires ffi.load which so file?

**Reply by chronolaw** [reply]
It is a posix api, need not to load any so file.
See: https://linux.die.net/man/3/clock_gettime

**Comment by chronolaw** [comment]
I have no test data about it, but I think that it is the only way to get microsecond unit.
